web-common/src/main/java/edu/unc/lib/boxc/web/common/services/AccessCopiesService.java

```diff
@@ -103,74 +109,99 @@ public boolean hasViewableFiles(ContentObjectRecord briefObj, AccessGroupSet pri
-        if (contentObj != null && permissionsHelper.hasOriginalAccess(principals, contentObj)) {
-            return contentObj;
+    public String getDownloadUrl(ContentObjectRecord contentObjectRecord, AccessGroupSet principals) {
+        if (contentObjectRecord == null || !permissionsHelper.hasOriginalAccess(principals, contentObjectRecord)) {
+            return "";
         }
-
-        return null;
+        return DatastreamUtil.getOriginalFileUrl(contentObjectRecord);
     }
 
+    private static final String IMAGE_CONTENT_TYPE = '^' + ContentCategory.image.getJoined();
+
     /**
-     * Get the path of the original_file datastream within contentObjectRecord that can be downloaded,
-     * or null if no appropriate original_file is present
      * @param contentObjectRecord
      * @param principals
-     * @return
+     * @param checkChildren if true, then in cases where it is ambiguous if the provided record has a thumbnail,
+     *             then additional queries will be performed to check.
+     * @return ID of the object the thumbnail for the provided object belongs to, or null if there is no thumbnail
      */
-    public String getDownloadUrl(ContentObjectRecord contentObjectRecord, AccessGroupSet principals) {
-        ContentObjectRecord contentObj = getContentObject(contentObjectRecord, principals);
-        if (contentObj != null) {
-            return DatastreamUtil.getOriginalFileUrl(contentObj);
+    public String getThumbnailId(ContentObjectRecord contentObjectRecord, AccessGroupSet principals,
+                                 boolean checkChildren) {
+        // Find thumbnail datastream recorded directly on the object, if present
+        var thumbId = DatastreamUtil.getThumbnailOwnerId(contentObjectRecord);
+        if (thumbId != null) {
+            log.debug("Found thumbnail object directly assigned to object {}", thumbId);
+            return thumbId;
+        }
+        // Don't need to check any further if object isn't a work or doesn't contain files with thumbnails
+        if (!ResourceType.Work.name().equals(contentObjectRecord.getResourceType())
+                || contentObjectRecord.getContentType() == null
+                || !contentObjectRecord.getContentType().contains(IMAGE_CONTENT_TYPE)) {
+            log.debug("Record {} is not applicable for a thumbnail", contentObjectRecord.getId());
+            return null;
+        }
+        if (!checkChildren) {
+            log.debug("Not checking children for work {}, so using self as thumbnail id", contentObjectRecord.getId());
+            return contentObjectRecord.getId();
+        }
+
+        SolrQuery query = buildFirstChildQuery(contentObjectRecord, principals);
+        // Limit query to just children which have a thumbnail datastream
+        query.addFilterQuery(solrSettings.getFieldName(SearchFieldKey.DATASTREAM.name()) + ":"
+                + DatastreamType.THUMBNAIL_LARGE.getId() + "|*");
```

Comment:
Is anything with a thumbnail always going to have a large thumbnail?

Reply:
I believe so, we don't have any workflows that produce just one size thumbnail. I went with the large size here since there were some existing parts where it was defaulting to large if the size wasn't specified, although it did mess up one test that was only populating small thumbnails in the datastream field